### ISSUE TITLE
Document border handling with padding

### ISSUE THREAD

**Issue by clemisch** [ASEM000/Kernex#12]
How does Kernex handle borders and out-of-bounds indices, especially with `padding="same"`? I skimmed the source code but did not find it. Very superficially it seems to me that it performs 0-padding. Is this the case? 

If yes, it would be useful to support different schemes like "mirror", "edge", or ideally dropping the out-of-bounds indices. 

Repro:

```python
from pylab import *
import jax.numpy as jnp
import kernex

@kernex.kmap(kernel_size=(11, 11), padding="same")
def kernel(patch):
    return jnp.mean(patch)

data = ones((100, 100))
out = kernel(data)

figure()
plot(out[50], "+-", label="Kernex")
plot(ones_like(out[50]), "+-", label="Ideal")
legend()
```
![image](https://github.com/ASEM000/Kernex/assets/5190547/d49e2aad-0141-4711-b728-9f0bf038e72e)

**Comment by ASEM000** [comment]
>  it seems to me that it performs 0-padding. Is this the case?

Yes, 

> If yes, it would be useful to support different schemes like "mirror", "edge", or ideally dropping the out-of-bounds indices.

I think it's simple to do this. I am thinking about something like this :

```python

@kex.kmap(pad_kwargs={"value":0,  ... })
... 
```
WDYT?

**Comment by clemisch** [comment]
Thanks, that looks good! Do I understand correctly that Kernex always uses array padding for border handling? 

Do you have experience with performance of padding vs. adapting the computed indices at the borders? If the latter is even possible with JAX and/or beneficial on accelerators.

FYI, it's possible to control how JAX handles out-of-bounds indices with [`Array.at[ind].get(mode=...)`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html). The default clips into the valid range, which should be analogous to `mode=edge` in `np.pad`.

**Comment by ASEM000** [comment]
Hello,

I extensively experimented with exploiting the mode parameter in .at (~jax 0.2) to avoid padding; surprisingly, the padding solution was faster on CPU/GPU for my setups.

How does this perform now with the new Jax versions?

**Comment by clemisch** [comment]
Ah, that's very interesting! 

I asked because I don't have good evidence myself. Naively, padding seems inefficient, but I don't know how it works under `jax.jit` and how the memory/flops tradeoff behaves especially on GPU. 

IIRC (very anecdotally) using `.at(mode="drop")` was noticably slower, maybe even x2 in my use cases. This was completely outside Kernex.

IMHO if padding is reasonably efficient and removes the need for custom border handling code, then it is the correct choice for Kernex :-) Exposing the `mode` would still be very handy though.

**Comment by clemisch** [comment]
PS: It could be interesting to compare performance of 

1. Kernex' `padding="same"` with array padding, to
2. no padding/border handling at all and `.at(mode="clip")` mimicking `pad(mode="edge")`.

If (2) is _significantly_ faster (and you deem this relevant) Kernex could use it as the default for `padding="same"`.

**Comment by ASEM000** [comment]
Ok, I will look into it and let you know.

**Comment by ASEM000** [comment]
@clemisch  feel free to reopen if you still have questions.

**Comment by clemisch** [comment]
Hey @ASEM000, thanks for the implementation!

I think at least `mode="reflect"` is buggy:

```python
import jax.numpy as jnp
import kernex

@kernex.kmap(
    kernel_size=(3,),
    padding=("same"),
    relative=False,
    padding_kwargs=dict(mode="reflect"),
)
def f(x):
    return x

x = jnp.array([1, 2, 3, 4, 5])
y = f(x)
z = jnp.pad(x, 1, mode="reflect")

print("x: ", x)
print("y: ", y)
print("z: ", z)
```

gives

```
x:  [1 2 3 4 5]
y:  [[3 1 2]     # <-- the `3` is incorrect, should be `2`
     [1 2 3]
     [2 3 4]
     [3 4 5]
     [4 5 4]]
z:  [2 1 2 3 4 5 4]   # <-- here, the first element is `2`
```

The Kernex output reflects incorrectly: the first element is `3` instead of `2`.

**Comment by clemisch** [comment]
FYI, I can't re-open.
